core/src/main/scala/akka/persistence/r2dbc/internal/BySliceQuery.scala

```diff
@@ -502,12 +519,23 @@ import org.slf4j.Logger
           .via(deserializeAndAddOffset(newState.currentOffset)))
     }
 
+    def heartbeat(state: QueryState): Option[Envelope] = {
+      if (state.idleCount >= 1 && state.latestBacktrackingWallClock != Instant.EPOCH) {
+        // using wall clock to measure duration since the last backtracking event
+        val timestamp =
+          state.latestBacktracking.timestamp.plus(JDuration.between(state.latestBacktrackingWallClock, clock.instant()))
+        createHeartbeat(timestamp)
```

Comment:
Ok, so rather than just wall clock time, we make this relative to what could be the db timestamp. Makes sense.

Reply:
So this can create some quite old timestamps. If replaying older events, then the latest backtracking timestamp can be from some time ago, while latest backtracking wall clock is from when that offset was seen. So whatever the distance is between those, we end up that far back from the current time.

Reply:
So I assume we wanted to avoid using the current wall clock instant, in case it's skewed from the database timestamps for events. But if we have a big enough gap between the event timestamp and when it was seen by backtracking, then the heartbeats will be outside the backtracking window and published events will need to be filtered. Won't resolve itself until new backtracking envelopes are seen, as before.

Reply:
Could be tricky to have a relative timestamp for all cases, as it needs to be from an event that was not originally written too long ago. With backtracking events, only works when it's up to date, and these are seen close enough to their write times. To have a more accurate marker for the latest db timestamp, maybe this could use published events instead — they'll be arriving soon after their writes. If we use the last seen published event + wall clock difference, then it could give an accurate relative heartbeat time.

Reply:
I guess basing it all in published events would need to be quite different. Mostly tracking it together in the pubsub event filtering, with just an empty heartbeat on idle coming from the query here. And maybe opens up the possibility of heartbeats actually getting ahead of later published events?

Maybe this should be quite simple. Get the current timestamp from the database when the query starts, and then just create relative timestamps from that start timestamp for each heartbeat on idle.

Reply:
Switched to using an initial database timestamp in #612.